include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -109,6 +110,191 @@ struct walk_n
     }
 };
 
+// Base class which establishes tuning parameters including vectorization / scalar path decider at compile time
+// for walk / for based algorithms
+template <typename... _Ranges>
+class walk_vector_or_scalar_base
+{
+    using _ValueTypes = std::tuple<oneapi::dpl::__internal::__value_t<_Ranges>...>;
+    constexpr static std::uint8_t __min_type_size = oneapi::dpl::__internal::__min_nested_type_size<_ValueTypes>::value;
+    // Empirically determined 'bytes-in-flight' to maximize bandwidth utilization
+    constexpr static std::uint8_t __bytes_per_item = 16;
+    // Maximum size supported by compilers to generate vector instructions
+    constexpr static std::uint8_t __max_vector_size = 4;
+
+  public:
+    constexpr static bool __can_vectorize =
+        (oneapi::dpl::__ranges::__is_vectorizable_range<_Ranges>::value && ...) &&
+        (std::is_fundamental_v<oneapi::dpl::__internal::__value_t<_Ranges>> && ...) && __min_type_size < 4;
+    // Vectorize for small types, so we generate 128-byte load / stores in a sub-group
+    constexpr static std::uint8_t __preferred_vector_size =
+        __can_vectorize ? oneapi::dpl::__internal::__dpl_ceiling_div(__max_vector_size, __min_type_size) : 1;
+    constexpr static std::uint8_t __preferred_iters_per_item =
+        __bytes_per_item / (__min_type_size * __preferred_vector_size);
+};
+
+// Path that intentionally disables vectorization for algorithms with a scattered access pattern (e.g. binary_search)
+template <typename... _Ranges>
+class walk_scalar_base
+{
+    using _ValueTypes = std::tuple<oneapi::dpl::__internal::__value_t<_Ranges>...>;
+    constexpr static std::uint8_t __min_type_size = oneapi::dpl::__internal::__min_nested_type_size<_ValueTypes>::value;
+    constexpr static std::uint8_t __bytes_per_item = 16;
+
+  public:
+    constexpr static bool __can_vectorize = false;
+    // With no vectorization, the vector size is 1
+    constexpr static std::uint8_t __preferred_vector_size = 1;
+    // To achieve full bandwidth utilization, multiple iterations need to be processed by a work item
+    constexpr static std::uint8_t __preferred_iters_per_item =
+        __bytes_per_item / (__min_type_size * __preferred_vector_size);
+};
+
+template <typename _ExecutionPolicy, typename _F, typename _Range>
+struct walk1_vector_or_scalar : public walk_vector_or_scalar_base<_Range>
+{
+    using __base_t = walk_vector_or_scalar_base<_Range>;
+    _F __f;
+    std::size_t __n;
+
+    template <typename _IsFull, typename _ItemId>
+    void
+    __vector_path(_IsFull __is_full, const _ItemId __idx, _Range __rng) const
+    {
+        // This is needed to enable vectorization
+        auto __raw_ptr = __rng.begin();
```

Comment:
1) I think that `__raw_ptr` isn't very good name because `begin()` usually linked in mind with iterator. But `raw` usually is some pointer.
2) Do we really need to have here local variable `__raw_ptr` ? Can we pass `__rng.begin()` instead of that variable into `__vector_walk` call?